cli/command_repository_upgrade.go

```diff
@@ -0,0 +1,240 @@
+	maxPermittedClockDrift time.Duration
+
+	svc advancedAppServices
+}
+
+func (c *commandRepositoryUpgrade) setup(svc advancedAppServices, parent commandParent) {
+	cmd := parent.Command("upgrade", "Upgrade repository format.")
+
+	// TODO:
+	// cmd.GetFlag("upgrade-owner-id").Required()
+	// cmd.("owner-id", "The upgrade lock unique owner ID").Required().StringVar(&c.ownerId)
+
+	cmd.Flag("advance-notice", "Advance notice for upgrade to allow enough time for other Kopia clients to notice the lock").DurationVar(&c.advanceNoticeInterval)
+	cmd.Flag("io-drain-timeout", "Max time it should take all other Kopia clients to drop repository connections").Default(repo.DefaultRepositoryBlobCacheDuration.String()).DurationVar(&c.ioDrainTimeout)
+	cmd.Flag("status-poll-interval", "An advisory polling interval to check for the status of upgrade").Default("60s").DurationVar(&c.statusPollInterval)
+	cmd.Flag("max-clock-drift", "Maximum tolerated drift on clocks between all Kopia clients").Default("5m").DurationVar(&c.maxPermittedClockDrift)
+	cmd.Flag("block-until-drain", "Drain all the clients but do not perform the upgrade").BoolVar(&c.blockUntilDrain)
+
+	// TODO(small): make rollback a sub-command instead of a flag
+	cmd.Flag("force-rollback", "Force rollback the repository upgrade, this action can cause repository corruption").BoolVar(&c.forceRollback)
+
+	// upgrade sequence
+	cmd.Action(svc.directRepositoryWriteAction(c.runPhase(c.setLockIntent)))
+	cmd.Action(svc.directRepositoryWriteAction(c.runPhase(c.drainOrCommit)))
+	cmd.Action(svc.directRepositoryWriteAction(c.runPhase(c.upgrade)))
+	cmd.Action(svc.directRepositoryWriteAction(c.runPhase(c.commitUpgrade)))
+
+	c.svc = svc
+}
+
+func (c *commandRepositoryUpgrade) runPhase(act func(context.Context, repo.DirectRepositoryWriter) error) func(context.Context, repo.DirectRepositoryWriter) error {
+	return func(ctx context.Context, rep repo.DirectRepositoryWriter) error {
+		if !c.skip {
+			if err := act(ctx, rep); err != nil {
+				// exc=plicitly skip all stages on error because tests do not
+				// skip/exit on error because they override os.Exit()
+				c.skip = true
+				return err
+			}
+		}
+
+		return nil
+	}
+}
+
+func (c *commandRepositoryUpgrade) setLockIntent(ctx context.Context, rep repo.DirectRepositoryWriter) error {
+	if c.forceRollback {
+		if err := rep.RollbackUpgrade(ctx); err != nil {
+			return errors.Wrap(err, "failed to rollback the upgrade")
+		}
+
+		c.skip = true
+
+		return nil
+	}
+
+	now := rep.Time()
+	mp := rep.ContentReader().ContentFormat().MutableParameters
+	openOpts := c.svc.optionsFromFlags(ctx)
+	l := &content.UpgradeLock{
+		OwnerID:                openOpts.UpgradeOwnerID,
+		CreationTime:           now,
+		AdvanceNoticeDuration:  c.advanceNoticeInterval,
+		IODrainTimeout:         c.ioDrainTimeout,
+		StatusPollInterval:     c.statusPollInterval,
+		Message:                fmt.Sprintf("Upgrading from format version %d -> %d", mp.Version, content.FormatVersion3),
```

Comment:
Is this always true?
```suggestion
		Message:                fmt.Sprintf("Upgrading from format version %d -> %d", mp.Version, content.MaxFormatVersion),
```

Reply:
Yes, that is true as of now.